src/relay/pass/pointer_analysis.cc

```diff
@@ -0,0 +1,168 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2019 by Contributors
+ *
+ * \file partial_eval.cc
+ *
+ * \brief Pointer analysis via Andersen's Algorithm.
+ * Analyze may-alias, escape, pointer content.
+ */
+#include <tvm/relay/analysis.h>
+#include <tvm/relay/expr_functor.h>
+#include "pass_util.h"
+
+namespace tvm {
+namespace relay {
+
+struct AllocAbstractLocation : ExprVisitor {
+  Map<Expr, AbstractLocation> spawn;
+  Map<AbstractLocation, Expr> origin;
+  void VisitExpr_(const RefCreateNode* op) final {
+    Expr refc = GetRef<RefCreate>(op);
+    auto loc = AbstractLocationNode::make(spawn.size() + 1);  // 0 is reserved for external ref
+    spawn.Set(refc, loc);
+    origin.Set(loc, refc);
+    ExprVisitor::VisitExpr_(op);
+  }
+};
+
+AbstractLocation external_ref = AbstractLocationNode::make(0);
+
+struct Analyze : ExprFunctor<void(const Expr&)> {
+  Map<Expr, AbstractLocation> spawn;
+  Map<AbstractLocation, Expr> origin;
+  Map<Expr, Set<AbstractLocation>> contain;
+  Map<AbstractLocation, Set<Expr>> store;
+  bool progress = true;
+  Analyze(Map<Expr, AbstractLocation> spawn,
+          Map<AbstractLocation, Expr> origin) : spawn(spawn), origin(origin) {
+    for (const auto& p : origin) {
+      store.Set(p.first, Set<Expr>());
+    }
+  }
+
+  void analyze(const Expr& e) {
+    progress = false;
+    VisitExpr(e);
+  }
+
+  void VisitExpr(const Expr& e) final {
+    if (contain.count(e) == 0) {
+      contain.Set(e, Set<AbstractLocation>());
+    }
+    ExprFunctor::VisitExpr(e);
+  }
+
+  void VisitExpr_(const ConstantNode* op) final { }
+  void VisitExpr_(const OpNode* op) final { }
+  void VisitExpr_(const ConstructorNode* op) final { }
+
+  void AddContain(const Expr& e, AbstractLocation loc) {
+    Set<AbstractLocation> locs = contain.at(e);
+    if (locs.count(loc) == 0) {
+      progress = true;
+      locs.Insert(loc);
+      contain.Set(e, locs);
+    }
+  }
+
+  void AddStore(AbstractLocation loc, const Expr& expr) {
+    Set<Expr> exprs = store.at(loc);
+    if (exprs.count(expr) == 0) {
+      progress = true;
+      exprs.Insert(expr);
+      store.Set(loc, exprs);
+    }
+  }
+
+  void UnionContain(const Expr& parent, const Expr& child) {
+    for (auto loc : contain.at(child)) {
+      AddContain(parent, loc);
+    }
+  }
```

Comment:
Maybe change the name of this function to indicate the union is asymmetric (i.e., only affecting `parent`).  `AddChildContain` or `AsymUnionContain` seem fine to me, if you don't think they're too verbose.